File: Hash Table/CourseSchedule.js
/**
 * Problem: https://leetcode.com/problems/course-schedule/submissions/
 * Author: Dung Nguyen Tien (Chris)
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    var learnt = Array(numCourses).fill(false);
    var total = 0;
    var map = createMap(numCourses, prerequisites);
    
    // if the course is not equal to total, continue learning
    while(total < numCourses) {
        var check = false;
        for (var c in map) {
            if (!learnt[c] && canLearn(learnt, map[c])) {
                learnt[c] = true;
                total++;
                check = true;
            }
        }
        
        // if go thorugh each course, cant learn anything, and the total is not the numCourses, mean having circle
        if (!check) return false;
    }
    return true;
};

// check if can learn this course
function canLearn(learnt, prerequisites) {
    var can = true;

    for (var p of prerequisites) can = can && learnt[p];
    
    return can;
}
         
// for each course, create the list of prerequisites
function createMap(numCourses, prerequisites) {
    var map = {};
    
    for (var i = 0; i < numCourses; i++) {
        map[i] = [];
    }
    
    for (var p of prerequisites) {
        map[p[0]].push(p[1]);
    }
    
    return map;
}

File: Topology Sort/CourseSchedule.cs
/*
 * Problem: https://leetcode.com/problems/course-schedule/submissions/
 * Author: Dung Nguyen Tien (Chris)
*/

public class Solution {
    // Topology Sort using BFS
    public bool CanFinish(int numCourses, int[][] prerequisites) {
        map = BuildMap(numCourses, prerequisites);
        var check = TopologySort(numCourses);
        Console.WriteLine(check);
        
        return check;
    }
    
    public Dictionary<int, List<int>> map;
    public int[] degree;
    public Dictionary<int, List<int>> BuildMap(int numCourses, int[][] prerequisites) {
        map = new Dictionary<int, List<int>>();
        degree = new int[numCourses];
        Array.Fill(degree, 0);
        
        for (var i = 0; i < numCourses; i++) map.Add(i, new List<int>());
        
        foreach(var item in prerequisites) {
            var pre = item[1];
            var course = item[0];
            
            map[pre].Add(course);
            degree[course]++;
        }
        return map;
    }
    
    public bool TopologySort(int numCourses) {
        var queue = new Queue<int>();
        var count = 0;
        for (var i = 0; i < numCourses; i++) {
            if (degree[i] == 0) queue.Enqueue(i);
        }
        
        while(queue.Count > 0) {
            var u = queue.Dequeue();
            
            foreach(var next in map[u]) {
                if(--degree[next] == 0) queue.Enqueue(next);
            }
            
            count++;
        }
        
        return count == numCourses;
    }
}